include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -202,103 +202,132 @@ struct __init_processing
+    template <typename _Size, typename _Res, typename... _Acc>
     void
-    seq_impl(const _NDItemId __item_id, const _Size __n, const _Size __global_offset, _Res& __res,
-             const _Acc&... __acc) const
+    vectorized_reduction_first(const _Size __start_idx, _Res& __res, const _Acc&... __acc) const
     {
-        auto __global_idx = __item_id.get_global_id(0);
-        auto __local_idx = __item_id.get_local_id(0);
-        const _Size __adjusted_global_id = __global_offset + __iters_per_work_item * __global_idx;
-        const _Size __adjusted_n = __global_offset + __n;
-        // Sequential load and reduce from global memory
-        if (__adjusted_global_id + __iters_per_work_item < __adjusted_n)
-        {
-            // Keep these statements in the same scope to allow for better memory alignment
-            new (&__res.__v) _Tp(__unary_op(__adjusted_global_id, __acc...));
-            _ONEDPL_PRAGMA_UNROLL
-            for (_Size __i = 1; __i < __iters_per_work_item; ++__i)
-                __res.__v = __binary_op(__res.__v, __unary_op(__adjusted_global_id + __i, __acc...));
-        }
-        else if (__adjusted_global_id < __adjusted_n)
-        {
-            const _Size __items_to_process = __adjusted_n - __adjusted_global_id;
-            // Keep these statements in the same scope to allow for better memory alignment
-            new (&__res.__v) _Tp(__unary_op(__adjusted_global_id, __acc...));
-            for (_Size __i = 1; __i < __items_to_process; ++__i)
-                __res.__v = __binary_op(__res.__v, __unary_op(__adjusted_global_id + __i, __acc...));
-        }
-        return;
+        new (&__res.__v) _Tp(__unary_op(__start_idx, __acc...));
+        _ONEDPL_PRAGMA_UNROLL
+        for (_Size __i = 1; __i < _VecSize; ++__i)
+            __res.__v = __binary_op(__res.__v, __unary_op(__start_idx + __i, __acc...));
     }
 
-    template <typename _NDItemId, typename _Size, typename _Res, typename... _Acc>
+    template <typename _Size, typename _Res, typename... _Acc>
     void
-    nonseq_impl(const _NDItemId __item_id, const _Size __n, const _Size __global_offset, _Res& __res,
-                const _Acc&... __acc) const
+    vectorized_reduction_remainder(const _Size __start_idx, _Res& __res, const _Acc&... __acc) const
     {
-        auto __local_idx = __item_id.get_local_id(0);
-        const _Size __stride = __item_id.get_local_range(0);
+        _ONEDPL_PRAGMA_UNROLL
+        for (_Size __i = 0; __i < _VecSize; ++__i)
+            __res.__v = __binary_op(__res.__v, __unary_op(__start_idx + __i, __acc...));
+    }
+
+    template <typename _Size, typename _Res, typename... _Acc>
+    void
+    scalar_reduction_remainder(const _Size __start_idx, const _Size __adjusted_n, const _Size __max_iters, _Res& __res,
+                               const _Acc&... __acc) const
+    {
+        const _Size __no_iters = std::min(static_cast<_Size>(__adjusted_n - __start_idx), __max_iters);
```

Comment:
`static_cast<_Size>` here and everywhere looks very suspicious and bulky...
Probably, we can pass a "right" integer type as _Size? Or/and use `auto` when it is applicable and doesn't break correctness?